Add PatternFly summary counts to security metrics dashboard.

dvail opened the pull request:
## Description

Adds a PF counterpart to the main dashboard summary counts, based on [MVP mocks](https://www.sketch.com/s/9d082d82-3515-40f2-875a-1b322a02bace/a/MyRkjGb).

## Checklist
- [ ] Investigated and inspected CI test results

## Testing Performed

Test that summary counts are displayed at the top of the new dashboard:
![image](https://user-images.githubusercontent.com/1292638/171725841-b69447ed-f98d-4888-bbc4-038906cf6168.png)

Test that a loading [skeleton](https://www.patternfly.org/v4/components/skeleton) is displayed in place of counts until the request completes (may require an artificial network tools slowdown):
![image](https://user-images.githubusercontent.com/1292638/171725971-718ec6ca-ad4e-4e04-9762-371028fba1b1.png)

Test that on error, an inline alert is shown instead of invalid data (may also require an artificial network tools failure):
![image](https://user-images.githubusercontent.com/1292638/171726186-c5fe41f8-eed7-4c01-ad58-250a829b3a1b.png)

Test that the new widget tolerates small screen sizes:
<img src="https://user-images.githubusercontent.com/1292638/171726292-1e76f8a4-b258-49bf-afa5-623560be9d3b.png" height=600>

dvail commented on ui/apps/platform/src/Containers/Dashboard/PatternFly/DashboardPage.css:
```diff
@@ -0,0 +1,13 @@
+/*
+These styles are to unset the conflicting rules from selector:
+#main-page-container > div > :not(.pf-c-page__main-section)
```
The referenced rule from `style.css` essentially overrides anything added to the main page component that isn't directly wrapped in a PF `<PageSection>`. I believe these are global styles for older Tailwind pages, but they also overwrite defaults on newer pages depending on the component used.

We might be able to change (or append) to that rule to make sure it is ignored for all PF pages by changing the selector to reference a CSS class from PF's `<Page>` component, and ensuring all new pages are wrapped with it. Probably better to discuss in a team meeting though, as it could have some wide reaching impact in the app.

pedrottimark replied:
This seems like a reasonable temporary work-around to keep forward momentum.

The need to diverge from general principle that content on PatternFly page is contained in one or more `PageSection` elements confirmed the need to follow up on question about recent trend in its use.

Looking at it in a different way, what are pro and con to render `SummaryCounts` element below the main heading of the page? Putting it above the heading was a temporary compromise because it had been on the classic top bar. In PatternFly it looks out of place above the **Dashboard** heading.

dvail replied:
Good question - I kept it like this because that's how it was in the mocks, and how it has been in the past. We could run the question by UX though and see if it makes sense to move it.

The main con I see is that it will be a little different than what users expect, and the impression I got from Slack convos is that users _really_ like this information as-is.

pedrottimark replied:
In case it needs to stay at the top, then what do you think about custom style if needed for a `PageSection` to contain `SummaryCounts` element, instead of exception to page element hierarchy?

Can you double check whether or not my memory and understanding is correct: idiomatic PatternFly for `Page` element is main content in one or more `PageSection` elements?

dvail replied:
Hmm it is a little ambiguous in the [PatternFly docs](https://www.patternfly.org/v4/components/page/accessibility):
> Page layouts are defined using page sections, such as the header, body, and footer of a page. 
but then also
> A page can be combined with many other components
, but I'd say leaning towards using <PageSection> in general.

I went with a non-section element due to the semantic issue you brought up the other day, but for consistency and avoiding this override I can convert to <PageSection>. From what I can see in our app, the vast majority (all?) of pages used either PageSection or Divider as the child elements of <Page>.

I also opened an [issue with PF](https://github.com/patternfly/patternfly-react/issues/7419) regarding switching out the element, so if they are in agreement we might be able to change these out with <div>s in the future.

dvail commented on ui/apps/platform/src/Containers/Dashboard/PatternFly/SummaryCounts.tsx:
```diff
@@ -0,0 +1,76 @@
+/*
+  TODO
+  We have a `TileContent` component that was used in the old version of this
+  widget at ./apps/platform/src/Components/TileContent/TileContent.tsx
```
The widget referenced here is used in a handful of pages that are on the old style. e.g.) Old main dashboard SummaryCounts, VulnMgmt -> Deployments -> Sidebar

I wasn't sure if this is a pattern we want to retain in PF, or if it is going to be slowly replaced as we migrate old pages. I created this Tile as a one-off for the summaries instead of porting the old Tile but can do that instead if it is something we need to use going forward.

dvail commented on ui/apps/platform/src/Containers/Dashboard/PatternFly/SummaryCounts.tsx:
```diff
@@ -0,0 +1,76 @@
+    if (isLoading) {
+        content = <Spinner size="lg" aria-label={`Total ${label} count`} />;
+    } else if (isError) {
+        content = <p title={`Failed to load ${label} counts`}>- -</p>;
```
The `- -` text here hasn't been run by UXD yet, but I wanted to something to cover the error case in the PR.

pedrottimark replied:
What are pro and con to lift up the loading and error states from `SummaryTile` to `SummaryCounts` because all of the tiles depend on one request?
* `if (loading) { return …; }`
* `if (error) { return …; }`
* `return …;`

dvail replied:
That was how I initially approached this, but I moved the loading states to the `SummaryTile` so that I could display loading and error states per tile. Moving these up to `SummaryCounts` would make the code a bit simpler, but would require an all-or-nothing widget to display loading and error states for the entirety of the tiles. Another question for UX - I don't think any of the mocks cover the loading/error states.

pedrottimark replied:
In this case, the loading and error state is all-or-nothing from summary counts request, true?

dvail replied:
Yes that is true, the loading/error state will always be identical, at the same time, for each individual tile.

dvail replied:
Hmm good question. Gut feeling is that if this is the case, the individual fields would need to be a union type with `number` and something else, but I'll see if I can find out for sure.

dvail replied:
Ok so it looks like in-theory we could get partial errors with this query, but the implementation sends a single error back if any part fails out:

```graphql
query q {
	clusterCount
	nodeCount
	violationCount
	deploymentCount
	imageCount(query: "Last Updated:>a")
	secretCount
}
```
returns a response of 
```json
{
	"errors": [
		{
			"message": "converting to bleve query: Invalid time query. Must be of the format (01/02/2006 or 1d)",
			"path": [
				"imageCount"
			]
		}
	],
	"data": null
}
```

dvail commented on ui/apps/platform/src/queries/summaryCounts.ts:
```diff
@@ -0,0 +1,21 @@
+import { gql } from '@apollo/client';
```
I moved this query and return type here, since it is being used by both the old and new dashboards at the same time. It seems that `src/queries` doesn't see a lot of updates though, is this the best location for shared queries?

(Alternatively, do we want to move away from GQL and just use the REST API here?)

dvail commented:
After talking with Mansur, I made the following updates to the UI:

- Changed the "loading" state to be a single skeleton component instead of individual spinners
- Changed the "error" state to be a PF "warning" alert
- Update the individual counters to link to the pages specified in the requirements doc, using the Link Button styling

These changes also helped clean up some of the complexity in the code, making is unnecessary to deal with the loading/error state on each individual tile.

pedrottimark commented on ui/apps/platform/src/Containers/Dashboard/PatternFly/SummaryCounts.tsx:
```diff
@@ -0,0 +1,104 @@
+        );
+    }
+
+    const tileData = {
```
What do you think about `Record<TileEntity, number>` as type for consistency with `tileLinks` above?

dvail replied:
Ah good catch. That was half the reason I created the TileEntity type. 😅
